### PR TITLE
Compiler now requires .NET 3.5 (compatible with Unity "old" runtime)

### PR DESCRIPTION
**Pull request by KumoKairo**
Relates to https://github.com/inkle/ink-unity-integration/issues/65
Also removed redundant "internal" access modifiers as they are harmful for 2017.3+ Assembly Definition files in Unity.

First step in making it work directly from Unity without a separate precompiled CLI program

Behavior is the same, all compilation tests pass.

### CODE REVIEW

**Comment by joethephish** [comment]
Thanks for this! So the small/main changes (collection tweaks, use of ToArray when joining etc) is all totally fine.

However I really don't want to open up all the internal stuff to be public though. The aim is to make the public facing footprint of the API as small as possible. So, ideally just the basic stuff within the `Compiler` class and nothing else. Sometimes it can be a little bit tricky to draw the line between public and internal but I'd much prefer to default to internal and then carefully carve out the public portion.

Is there a specific reason that you opened all that stuff up?

**Comment by joethephish** [comment]
Oh, also, it looks like you've renamed "ink-engine-runtime" to "ink-engine-runtime_3_5" and "ink_compiler" to "ink_compiler_3_5" - is that necessary?

**Comment by KumoKairo** [comment]
Oh, good catch on piblic/internal, I see your initial intentions now. I have initially thought all of this `internal` stuff would be needed only to reduce clashing of the source runtime in Unity and the built one in dll (the one that Compiler uses). 
I have marked it as public so it's possible to access Compiler API directly from C# scripts inside Unity even when it's build as a separate assembly with Assembly Definition files (or built as a .dll).
By default, however, Unity puts every script from the Script folder in the same assemblies, essentially making all `internal` things actually visible to all other Unity scripts in the project, rendering intentions of `internal` stuff somewhat useless (as both Runtime, Compiler and all other scripts in Unity project are placed in the same assembly, all `internal` stuff is visible everywhere, just like it would be if it was marked as public). The only real way of making something actually `internal` is to either:
1. Build it manually  and place a `dll` inside Unity's project folder
2. Putting different Ink-related folders in different Assemblies using Assembly Definitions (2017.3+)
In other cases, again, `internal` would actually be exactly like `public` inside Unity.

So I have just replaced `internal` with `public` because this is how runtime worked inside Unity anyway. If it's a problem, we need to decide how to solve it. It may be better to mark certain things as `private` instead, and provide a public Compiler API as public (i.e. `Compiler.Compile()` etc.)

As for the naming - I have initially named it like that because it was required to make separate projects with unique names. After making all changes, I have named it back to the original names:
https://github.com/KumoKairo/ink/blob/features/older_compiler_runtime/ink.sln#L12
https://github.com/KumoKairo/ink/blob/features/older_compiler_runtime/compiler/ink_compiler.csproj#L11
I will double check that it doesn't have that 3_5 naming anywhere, it shouldn't be named like that anywhere :)

**Comment by joethephish** [comment]
So my general approach to this (which is what I've done on the runtime side) is to use internal so that:

* If you want to use a .dll then the correct very minimal API footprint is exposed.
* But if you use source directly in Unity, then you can hack around and call non-public stuff if you want. However when looking at source, the internal keyword should convey the intent at the very least.

I'd also be happy to keep large portions entirely private (like everything on the parser side), though it's probably easier/more straightforward to default to internal since otherwise it'll be a lot of work defining yet another boundary between private/internal. And stuff like the `Parsed.Object` hierarchy should probably be internal so it's possible to optionally hack around with that structure if you want to (when including source).

I'm also trying to keep things "correct" from a C# point of view, so even if you're not using Unity then the assembly is exposing the right things.

Naming: ah great okay!

**Comment by KumoKairo** [comment]
Cool, I got your point. Will revisit internal / public stuff and at the very least make it internal as it was, but usable outside of the built dll so the compiler is accessible.

File: compiler/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -1,28 +1,36 @@
 ﻿using System.Reflection;
 using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
 
-// Information about this assembly is defined by the following attributes. 
-// Change them to the values specific to your project.
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("ink_compiler_3_5")]
```
[the remark this answers is not in the document]
Comment (joethephish) [reply compiler/Properties/AssemblyInfo.cs]: Is it necessary to change this file at all...?

[the remark this answers is not in the document]
Comment (KumoKairo) [reply compiler/Properties/AssemblyInfo.cs]: It's a completely different file, portable projects are not convertable to old .NET types. I have just changed it to match the old one (see latest commit)

[the remark this answers is not in the document]
Comment (KumoKairo) [reply compiler/Properties/AssemblyInfo.cs]: https://github.com/KumoKairo/ink/blob/features/older_compiler_runtime/compiler/Properties/AssemblyInfo.cs#L6

**Comment by KumoKairo** [comment]
It seems that this internal/protected thing drags itself into the hacky InternalsVisibleTo which allows to supress the accessibility restrictions, and right now it somehow doesn't allow me to mark the `ink-tests` assembly to be an "assembly friend" to the Runtime (works fine for `compiler` and `inklecate`), Will investigate further. 
However, in my opinion, these smelly "not visible to except" snippets just increase overall complexity and cognitive load on the code. But one thing to solve at a time :)

**Comment by joethephish** [comment]
TBH I don't have a problem with `InternalsVisibleTo` for a bunch of very closely related modules like these (compiler, runtime, tests etc), since it allows us to keep things modular and closed to the outside world while allowing flexibility internally.

**Comment by joethephish** [comment]
By the way, not sure whether you're in the middle of something but I'm currently finding that these project files have stopped working for me in Xamarin on Mac:

image

**Comment by KumoKairo** [comment]
Yes, the changes are transient (I'm "in the middle of something"), they are not production ready yet. We should probably close this PR for now so it's not misleading whether or not it's ready for merging.
Will reopen it once it's done

**Comment by KumoKairo** [comment]
I have created another solution for this, with as little intrusion as possible:
https://github.com/inkle/ink/compare/master...KumoKairo:features/3_5_runtime
The only thing left is to check the projects on Xamarin Mac and will create another pullrequest for this

**Pull request closed without merging**